src/gluonts/ev/api.py

```diff
@@ -0,0 +1,150 @@
+# permissions and limitations under the License.
+
+from dataclasses import dataclass
+from typing import Callable, Dict, Iterator, Collection, List, Optional
+import numpy as np
+
+from gluonts.model.forecast import Forecast
+from gluonts.dataset.split import TestData
+from gluonts.ev.helpers import EvalData, create_eval_data
+from gluonts.ev.aggregations import Aggregation
+
+
+def gather_inputs(
+    test_data: TestData,
+    forecasts: Iterator[Forecast],
+    quantile_levels: Collection[float],
+    batch_size: int = 64,
+) -> Iterator[EvalData]:
+    """Collect relevant data as NumPy arrays to evaluate the next batch.
+
+    The number of entries in `test_data` and `forecasts` must be equal.
+    """
+    inputs = iter(test_data.input)
+    labels = iter(test_data.label)
+
+    input_data = []
+    label_data = []
+    forecast_data = {
+        "mean": [],
+        **{str(q): [] for q in quantile_levels},
+    }
+
+    entry_counter = 0
+    for _ in range(len(test_data)):
+        forecast_entry = next(forecasts)
+        forecast_data["mean"].append(forecast_entry.mean)
+        for q in quantile_levels:
+            forecast_data[str(q)].append(forecast_entry.quantile(q))
+
+        input_data.append(next(inputs))
+        label_data.append(next(labels))
+
+        entry_counter += 1
+        if entry_counter % batch_size == 0:
+            yield create_eval_data(input_data, label_data, forecast_data)
+
+            input_data.clear()
+            label_data.clear()
+            for key in forecast_data:
+                forecast_data[key].clear()
+
+    if entry_counter % batch_size != 0:
+        yield create_eval_data(input_data, label_data, forecast_data)
+
+
+class DataProbe:
+    """A DataProbe gathers all quantile forecasts required for an evaluation.
+
+    This has the benefit that metric definitions can work independently of
+    `Forecast` objects as all values in 'data' will be NumPy arrays.
+    :raises ValueError: if a metric requests a key that can't be converted to
+        float and isn't equal to "batch_size", "input", "label" or "mean"
+    """
+
+    def __init__(self, test_data: TestData):
+        input_sample, label_sample = next(iter(test_data))
+        # use batch_size 1
+        self.input_shape = (1,) + np.shape(input_sample["target"])
+        self.prediction_target_shape = (1,) + np.shape(label_sample["target"])
+
+        self.required_quantile_forecasts = set()
+
+    def __getitem__(self, key: str):
+        if key == "batch_size":
+            return 1
+        if key == "input":
+            return np.random.rand(*self.input_shape)
+        if key in ["label", "mean"]:
+            return np.random.rand(*self.prediction_target_shape)
+
+        try:
+            self.required_quantile_forecasts.add(float(key))
+            return np.random.rand(*self.prediction_target_shape)
+        except ValueError:
+            raise ValueError(f"Unexpected input: {key}")
+
+
+class Metric:
```

Comment:
I think we can make this a protocol or type-alias.

Reply:
Now, metric is `Protocol`. Is it correct that we now don't need to inherit from `Metric` anymore (for example, turning `class MeanSquaredError(Metric):` into `class MeanSquaredError:`)?